File: apps/ios/Sources/BikeCore/WorkspaceMutations.swift
import Foundation

public extension Workspace {
    func withActiveDocument(_ documentId: String) -> Workspace {
        guard documents.contains(where: { $0.id == documentId }) else { return self }
        var next = self
        next.activeDocumentId = documentId
        return next
    }

    func withDocumentTitle(documentId: String, title: String, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.title = String(title.prefix(120))
        }
    }

    func withDocumentShortcut(documentId: String, isShortcut: Bool, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now, preserveMarkdown: true) { document in
            document.isShortcut = isShortcut
        }
    }

    func withDocumentMovedToFront(_ documentId: String) -> Workspace {
        guard let index = documents.firstIndex(where: { $0.id == documentId }), index > 0 else {
            return self
        }
        var nextDocuments = documents
        let document = nextDocuments.remove(at: index)
        nextDocuments.insert(document, at: 0)
        var next = self
        next.documents = nextDocuments
        return next
    }

    func withDocumentDuplicated(documentId: String, now: Date = Date()) -> Workspace {
        guard let index = documents.firstIndex(where: { $0.id == documentId }) else { return self }
        let duplicated = documents[index].duplicated(now: now)
        var next = self
        next.documents.insert(duplicated, at: index + 1)
        next.activeDocumentId = duplicated.id
        return next
    }

    func withDocumentsDuplicated(documentIds: Set<String>, now: Date = Date()) -> Workspace {
        guard !documentIds.isEmpty else { return self }
        var lastCopiedId: String?
        var nextDocuments: [OutlineDocument] = []
        for document in documents {
            nextDocuments.append(document)
            if documentIds.contains(document.id) {
                let copy = document.duplicated(now: now)
                nextDocuments.append(copy)
                lastCopiedId = copy.id
            }
        }
        var next = self
        next.documents = nextDocuments
        next.activeDocumentId = lastCopiedId ?? activeDocumentId
        return next
    }

    func withDocumentDeleted(documentId: String, now: Date = Date()) -> Workspace {
        withDocumentsDeleted(documentIds: [documentId], now: now)
    }

    func withDocumentsDeleted(documentIds: Set<String>, now: Date = Date()) -> Workspace {
        guard !documentIds.isEmpty else { return self }
        let remaining = documents.filter { !documentIds.contains($0.id) }
        if !remaining.isEmpty {
            var next = self
            next.documents = remaining
            if documentIds.contains(activeDocumentId) {
                next.activeDocumentId = remaining[0].id
            }
            return next
        }

        let timestamp = ISO8601DateFormatter.bike.string(from: now)
        let document = OutlineDocument(
            id: newBikeId("doc"),
            title: "未命名文档",
            createdAt: timestamp,
            updatedAt: timestamp,
            nodes: [outlineNode("新主题")]
        )
        return Workspace(activeDocumentId: document.id, documents: [document])
    }

    func withInboxEntry(_ content: String, now: Date = Date()) -> Workspace {
        let trimmed = content.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return self }

        let timestamp = ISO8601DateFormatter.bike.string(from: now)
        let lines = trimmed
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        let title = String((lines.first ?? "分享内容").prefix(120))
        let note = lines.dropFirst().joined(separator: "\n")
        let node = outlineNode(title, note: note)

        if let inboxIndex = documents.firstIndex(where: { $0.title == inboxDocumentTitle }) {
            var next = self
            next.activeDocumentId = next.documents[inboxIndex].id
            next.documents[inboxIndex].updatedAt = timestamp
            next.documents[inboxIndex].nodes.insert(node, at: 0)
            next.documents[inboxIndex].markdownSource = nil
            next.documents[inboxIndex].markdownUpdatedAt = nil
            return next
        }

        let document = OutlineDocument(
            id: newBikeId("doc"),
            title: inboxDocumentTitle,
            createdAt: timestamp,
            updatedAt: timestamp,
            nodes: [node]
        )
        var next = self
        next.activeDocumentId = document.id
        next.documents.insert(document, at: 0)
        return next
    }

    func withNodeText(documentId: String, nodeId: String, text: String, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.updateNode(nodeId) { node in
                node.text = text
            }
        }
    }

    func withNodeTextAndNote(documentId: String, nodeId: String, text: String, note: String, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.updateNode(nodeId) { node in
                node.text = text
                node.note = note
            }
        }
    }

    func withNodeChecked(documentId: String, nodeId: String, checked: Bool, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.updateNode(nodeId) { node in
                node.checked = checked
            }
        }
    }

    func withNodeCollapsed(documentId: String, nodeId: String, collapsed: Bool, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now, preserveMarkdown: true) { document in
            document.nodes = document.nodes.updateNode(nodeId) { node in
                node.collapsed = collapsed
            }
        }
    }

    func withSiblingAfter(documentId: String, nodeId: String, newNode: OutlineNode, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.insertSiblingAfter(nodeId, newNode: newNode)
        }
    }

    func withChildNode(documentId: String, nodeId: String, childNode: OutlineNode, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.addChild(nodeId, childNode: childNode)
        }
    }

    func withGeneratedOutlineChildren(documentId: String, nodeId: String, children: [OutlineNode], now: Date = Date()) -> Workspace {
        let validChildren = children.filter { !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        guard !validChildren.isEmpty else { return self }
        return updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.updateNode(nodeId) { node in
                node.collapsed = false
                node.children.append(contentsOf: validChildren)
            }
        }
    }

    func withNodeMovedToParentLevel(documentId: String, nodeId: String, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes = document.nodes.outdentNode(nodeId)
        }
    }

    func withRootNode(documentId: String, newNode: OutlineNode, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            document.nodes.append(newNode)
        }
    }

    func withNodeDeleted(documentId: String, nodeId: String, now: Date = Date()) -> Workspace {
        updateDocument(documentId, now: now) { document in
            let nextNodes = document.nodes.deleteNode(nodeId)
            document.nodes = nextNodes.isEmpty ? [outlineNode("新主题")] : nextNodes
        }
    }

    func withNewDocument(title: String = "未命名文档", now: Date = Date()) -> Workspace {
        let timestamp = ISO8601DateFormatter.bike.string(from: now)
        let document = OutlineDocument(
            id: newBikeId("doc"),
            title: title,
            createdAt: timestamp,
            updatedAt: timestamp,
            nodes: [outlineNode("新主题")]
        )
        var next = self
        next.activeDocumentId = document.id
        next.documents.insert(document, at: 0)
        return next
    }

    private func updateDocument(_ documentId: String, now: Date, preserveMarkdown: Bool = false, update: (inout OutlineDocument) -> Void) -> Workspace {
        var next = self
        guard let index = next.documents.firstIndex(where: { $0.id == documentId }) else {
            return self
        }
        update(&next.documents[index])
        guard next.documents[index] != documents[index] else { return self }
        if !preserveMarkdown {
            next.documents[index].markdownSource = nil
            next.documents[index].markdownUpdatedAt = nil
        }
        next.documents[index].updatedAt = ISO8601DateFormatter.bike.string(from: now)
        return next
    }
}

public extension OutlineDocument {
    func duplicated(now: Date = Date()) -> OutlineDocument {
        let timestamp = ISO8601DateFormatter.bike.string(from: now)
        return OutlineDocument(
            id: newBikeId("doc"),
            title: title.isEmpty ? "副本" : String("\(title) 副本".prefix(120)),
            createdAt: timestamp,
            updatedAt: timestamp,
            isShortcut: isShortcut,
            nodes: nodes.map { $0.duplicated() }
        )
    }
}

public extension OutlineNode {
    func duplicated() -> OutlineNode {
        var next = self
        next.id = newBikeId("node")
        next.children = children.map { $0.duplicated() }
        return next
    }
}

public extension Array where Element == OutlineNode {
    func findNode(_ nodeId: String) -> OutlineNode? {
        for node in self {
            if node.id == nodeId {
                return node
            }
            if let child = node.children.findNode(nodeId) {
                return child
            }
        }
        return nil
    }

    func updateNode(_ nodeId: String, update: (inout OutlineNode) -> Void) -> [OutlineNode] {
        map { node in
            var next = node
            if next.id == nodeId {
                update(&next)
            } else {
                next.children = next.children.updateNode(nodeId, update: update)
            }
            return next
        }
    }

    func insertSiblingAfter(_ nodeId: String, newNode: OutlineNode) -> [OutlineNode] {
        if let index = firstIndex(where: { $0.id == nodeId }) {
            var next = self
            next.insert(newNode, at: index + 1)
            return next
        }
        return map { node in
            var next = node
            next.children = next.children.insertSiblingAfter(nodeId, newNode: newNode)
            return next
        }
    }

    func addChild(_ nodeId: String, childNode: OutlineNode) -> [OutlineNode] {
        updateNode(nodeId) { node in
            node.collapsed = false
            node.children.append(childNode)
        }
    }

    func deleteNode(_ nodeId: String) -> [OutlineNode] {
        compactMap { node in
            if node.id == nodeId { return nil }
            var next = node
            next.children = next.children.deleteNode(nodeId)
            return next
        }
    }

    func outdentNode(_ nodeId: String) -> [OutlineNode] {
        for index in indices {
            if let childIndex = self[index].children.firstIndex(where: { $0.id == nodeId }) {
                var next = self
                let target = next[index].children.remove(at: childIndex)
                next.insert(target, at: index + 1)
                return next
            }
        }
        return map { node in
            var next = node
            next.children = next.children.outdentNode(nodeId)
            return next
        }
    }

    func flattenVisible(depth: Int = 0) -> [FlatNodeRow] {
        flatMap { node -> [FlatNodeRow] in
            let current = [FlatNodeRow(node: node, depth: depth)]
            return node.collapsed ? current : current + node.children.flattenVisible(depth: depth + 1)
        }
    }

    func flattenSearch(_ query: String, depth: Int = 0) -> [FlatNodeRow] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return flattenVisible(depth: depth) }
        return flatMap { node -> [FlatNodeRow] in
            let current = node.matches(trimmed) ? [FlatNodeRow(node: node, depth: depth)] : []
            return current + node.children.flattenSearch(trimmed, depth: depth + 1)
        }
    }
}

public extension OutlineNode {
    func matches(_ query: String) -> Bool {
        text.localizedCaseInsensitiveContains(query) || note.localizedCaseInsensitiveContains(query)
    }
}

File: apps/android/app/src/main/java/com/bike/android/data/WorkspaceMutations.kt
package com.bike.android.data

import java.time.Instant

fun Workspace.withActiveDocument(documentId: String): Workspace =
    if (documents.any { it.id == documentId }) copy(activeDocumentId = documentId) else this

fun Workspace.withDocumentTitle(
    documentId: String,
    title: String,
    now: Instant = Instant.now(),
): Workspace {
    val normalizedTitle = title.take(120)
    return updateDocument(documentId) { document ->
        document.copy(
            title = normalizedTitle,
            updatedAt = now.toString(),
        )
    }
}

fun Workspace.withDocumentShortcut(
    documentId: String,
    isShortcut: Boolean,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId, preserveMarkdown = true) { document ->
        document.copy(
            isShortcut = isShortcut,
            updatedAt = now.toString(),
        )
    }

fun Workspace.withDocumentMovedToFront(
    documentId: String,
): Workspace {
    val index = documents.indexOfFirst { it.id == documentId }
    if (index <= 0) return this

    val document = documents[index]
    return copy(
        documents = listOf(document) + documents.filterNot { it.id == documentId },
    )
}

fun Workspace.withDocumentDuplicated(
    documentId: String,
    now: Instant = Instant.now(),
): Workspace {
    val source = documents.firstOrNull { it.id == documentId } ?: return this
    val duplicatedDocument = source.duplicated(now)
    val sourceIndex = documents.indexOfFirst { it.id == documentId }
    val nextDocuments = documents.toMutableList().apply {
        add(sourceIndex + 1, duplicatedDocument)
    }
    return copy(
        activeDocumentId = duplicatedDocument.id,
        documents = nextDocuments,
    )
}

fun Workspace.withDocumentsDuplicated(
    documentIds: Set<String>,
    now: Instant = Instant.now(),
): Workspace {
    if (documentIds.isEmpty()) return this
    var lastCopiedId: String? = null
    val nextDocuments = documents.flatMap { document ->
        if (document.id in documentIds) {
            val copy = document.duplicated(now)
            lastCopiedId = copy.id
            listOf(document, copy)
        } else {
            listOf(document)
        }
    }
    return copy(
        activeDocumentId = lastCopiedId ?: activeDocumentId,
        documents = nextDocuments,
    )
}

fun Workspace.withDocumentDeleted(
    documentId: String,
    now: Instant = Instant.now(),
): Workspace =
    withDocumentsDeleted(setOf(documentId), now)

fun Workspace.withDocumentsDeleted(
    documentIds: Set<String>,
    now: Instant = Instant.now(),
): Workspace {
    if (documentIds.isEmpty()) return this

    val remaining = documents.filterNot { it.id in documentIds }
    if (remaining.isNotEmpty()) {
        val nextActiveId = if (activeDocumentId in documentIds) {
            remaining.first().id
        } else {
            activeDocumentId
        }
        return copy(
            activeDocumentId = nextActiveId,
            documents = remaining,
        )
    }

    val timestamp = now.toString()
    val document = OutlineDocument(
        id = newBikeId("doc"),
        title = "未命名文档",
        createdAt = timestamp,
        updatedAt = timestamp,
        nodes = listOf(outlineNode("新主题")),
    )
    return copy(
        activeDocumentId = document.id,
        documents = listOf(document),
    )
}

fun Workspace.withNodeChecked(
    documentId: String,
    nodeId: String,
    checked: Boolean,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        val updatedNodes = document.nodes.updateNode(nodeId) { node ->
            node.copy(checked = checked)
        }
        document.copy(nodes = updatedNodes, updatedAt = now.toString())
    }

fun Workspace.withInboxEntry(
    content: String,
    now: Instant = Instant.now(),
): Workspace {
    val trimmed = content.trim()
    if (trimmed.isBlank()) return this

    val timestamp = now.toString()
    val lines = trimmed.lines().map { it.trim() }.filter { it.isNotBlank() }
    val title = lines.firstOrNull()?.take(120) ?: "分享内容"
    val note = lines.drop(1).joinToString("\n")
    val node = outlineNode(text = title, note = note)
    val inbox = documents.firstOrNull { it.title == INBOX_DOCUMENT_TITLE }

    if (inbox == null) {
        val document = OutlineDocument(
            id = newBikeId("doc"),
            title = INBOX_DOCUMENT_TITLE,
            createdAt = timestamp,
            updatedAt = timestamp,
            nodes = listOf(node),
        )
        return copy(
            activeDocumentId = document.id,
            documents = listOf(document) + documents,
        )
    }

    return copy(
        activeDocumentId = inbox.id,
        documents = documents.map { document ->
            if (document.id == inbox.id) {
                document.copy(
                    updatedAt = timestamp,
                    nodes = listOf(node) + document.nodes,
                    markdownSource = null,
                    markdownUpdatedAt = null,
                )
            } else {
                document
            }
        },
    )
}

fun Workspace.withNodeText(
    documentId: String,
    nodeId: String,
    text: String,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(text = text)
            },
            updatedAt = now.toString(),
        )
    }

fun Workspace.withNodeNote(
    documentId: String,
    nodeId: String,
    note: String,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(note = note)
            },
            updatedAt = now.toString(),
        )
    }

fun Workspace.withNodeTextAndNote(
    documentId: String,
    nodeId: String,
    text: String,
    note: String,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(text = text, note = note)
            },
            updatedAt = now.toString(),
        )
    }

fun Workspace.withNodeCollapsed(
    documentId: String,
    nodeId: String,
    collapsed: Boolean,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId, preserveMarkdown = true) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(collapsed = collapsed)
            },
            updatedAt = now.toString(),
        )
    }

fun Workspace.withSiblingAfter(
    documentId: String,
    nodeId: String,
    text: String = "新主题",
    now: Instant = Instant.now(),
): Workspace =
    withSiblingAfter(
        documentId = documentId,
        nodeId = nodeId,
        newNode = outlineNode(text),
        now = now,
    )

fun Workspace.withSiblingAfter(
    documentId: String,
    nodeId: String,
    newNode: OutlineNode,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.insertSiblingAfter(nodeId, newNode),
            updatedAt = now.toString(),
        )
    }

fun Workspace.withChildNode(
    documentId: String,
    nodeId: String,
    text: String = "新子主题",
    now: Instant = Instant.now(),
): Workspace =
    withChildNode(
        documentId = documentId,
        nodeId = nodeId,
        childNode = outlineNode(text),
        now = now,
    )

fun Workspace.withChildNode(
    documentId: String,
    nodeId: String,
    childNode: OutlineNode,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(
                    collapsed = false,
                    children = node.children + childNode,
                )
            },
            updatedAt = now.toString(),
        )
    }

fun Workspace.withGeneratedChildren(
    documentId: String,
    nodeId: String,
    texts: List<String>,
    now: Instant = Instant.now(),
): Workspace {
    val children = texts
        .map { it.trim() }
        .filter { it.isNotBlank() }
        .map { outlineNode(it) }
    if (children.isEmpty()) return this

    return updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(
                    collapsed = false,
                    children = node.children + children,
                )
            },
            updatedAt = now.toString(),
        )
    }
}

fun Workspace.withGeneratedOutlineChildren(
    documentId: String,
    nodeId: String,
    children: List<OutlineNode>,
    now: Instant = Instant.now(),
): Workspace {
    val validChildren = children.filter { it.text.isNotBlank() }
    if (validChildren.isEmpty()) return this

    return updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.updateNode(nodeId) { node ->
                node.copy(
                    collapsed = false,
                    children = node.children + validChildren,
                )
            },
            updatedAt = now.toString(),
        )
    }
}

fun Workspace.withNodeMovedToParentLevel(
    documentId: String,
    nodeId: String,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes.outdentNode(nodeId),
            updatedAt = now.toString(),
        )
    }

fun Workspace.withRootNode(
    documentId: String,
    text: String = "新主题",
    now: Instant = Instant.now(),
): Workspace =
    withRootNode(
        documentId = documentId,
        newNode = outlineNode(text),
        now = now,
    )

fun Workspace.withRootNode(
    documentId: String,
    newNode: OutlineNode,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        document.copy(
            nodes = document.nodes + newNode,
            updatedAt = now.toString(),
        )
    }

fun Workspace.withNodeDeleted(
    documentId: String,
    nodeId: String,
    now: Instant = Instant.now(),
): Workspace =
    updateDocument(documentId) { document ->
        val nextNodes = document.nodes.deleteNode(nodeId)
        document.copy(
            nodes = nextNodes.ifEmpty { listOf(outlineNode("新主题")) },
            updatedAt = now.toString(),
        )
    }

fun Workspace.withNewDocument(
    title: String = "未命名文档",
    now: Instant = Instant.now(),
): Workspace {
    val timestamp = now.toString()
    val document = OutlineDocument(
        id = newBikeId("doc"),
        title = title,
        createdAt = timestamp,
        updatedAt = timestamp,
        nodes = listOf(outlineNode("新主题")),
    )

    return copy(
        activeDocumentId = document.id,
        documents = listOf(document) + documents,
    )
}

private fun OutlineDocument.duplicated(now: Instant): OutlineDocument {
    val timestamp = now.toString()
    return copy(
        id = newBikeId("doc"),
        title = title.copyTitle(),
        createdAt = timestamp,
        updatedAt = timestamp,
        markdownSource = null,
        markdownUpdatedAt = null,
        nodes = nodes.map { it.duplicated() },
    )
}

private fun OutlineNode.duplicated(): OutlineNode =
    copy(
        id = newBikeId("node"),
        children = children.map { it.duplicated() },
    )

private fun String.copyTitle(): String =
    if (isBlank()) {
        "副本"
    } else {
        "$this 副本".take(120)
    }

private fun Workspace.updateDocument(
    documentId: String,
    preserveMarkdown: Boolean = false,
    update: (OutlineDocument) -> OutlineDocument,
): Workspace =
    copy(
        documents = documents.map { document ->
            if (document.id == documentId) {
                val updated = update(document)
                if (preserveMarkdown || updated.copy(updatedAt = document.updatedAt) == document) updated
                else updated.copy(markdownSource = null, markdownUpdatedAt = null)
            } else document
        },
    )

private fun List<OutlineNode>.updateNode(
    nodeId: String,
    update: (OutlineNode) -> OutlineNode,
): List<OutlineNode> =
    map { node ->
        when {
            node.id == nodeId -> update(node)
            node.children.isNotEmpty() -> node.copy(children = node.children.updateNode(nodeId, update))
            else -> node
        }
    }

private fun List<OutlineNode>.insertSiblingAfter(
    nodeId: String,
    newNode: OutlineNode,
): List<OutlineNode> {
    val directIndex = indexOfFirst { it.id == nodeId }
    if (directIndex >= 0) {
        return toMutableList().apply {
            add(directIndex + 1, newNode)
        }
    }

    return map { node ->
        if (node.children.isEmpty()) {
            node
        } else {
            node.copy(children = node.children.insertSiblingAfter(nodeId, newNode))
        }
    }
}

private fun List<OutlineNode>.deleteNode(nodeId: String): List<OutlineNode> =
    mapNotNull { node ->
        when {
            node.id == nodeId -> null
            node.children.isNotEmpty() -> node.copy(children = node.children.deleteNode(nodeId))
            else -> node
        }
    }

private fun List<OutlineNode>.outdentNode(nodeId: String): List<OutlineNode> {
    forEachIndexed { index, node ->
        val childIndex = node.children.indexOfFirst { it.id == nodeId }
        if (childIndex >= 0) {
            val target = node.children[childIndex]
            val nextParent = node.copy(
                children = node.children.toMutableList().apply {
                    removeAt(childIndex)
                },
            )
            return toMutableList().apply {
                set(index, nextParent)
                add(index + 1, target)
            }
        }
    }

    forEachIndexed { index, node ->
        if (node.children.isNotEmpty()) {
            val nextChildren = node.children.outdentNode(nodeId)
            if (nextChildren !== node.children) {
                return toMutableList().apply {
                    set(index, node.copy(children = nextChildren))
                }
            }
        }
    }

    return this
}
